Version 3.2.0

UselessGuru opened the pull request:
## Changelog Version 3.2.0

#### New features
- Added parameter '-PoolBalancesUpdateInterval'

- Pool fee is now part of the profitability calculation

- Separate section 'Wallets' in config file, BTC/Wallet gets automatically moved to 'Wallets' on start

#### Core changes
- Added 'Bitcoin' to the list of unprofitable algos
- Added EquihashPers.txt to store Equihash 'pers' values
- Added 'poolbalancesdata' to web API

- Changed APIs files; simplified algo name detection
- Changed defaults parameter values for fresh installations; enabled 'CreateMinerInstancePerDeviceModel' and 'UseDeviceNameForStatsFileNaming' in config file
- Changed exchange rate currency name to be displayed in upper case
- Changed list of default algos in start batch files:
  - Start-MiningPoolHub: removed lyra2re2
  - Start_Zpool: aded hex,x16r,x16s,x17,x22i; removed lyra2re2

- Enhancement: Better detection of failed miners (some crashed miners are dead, but may still be found by their processId)
- Enhancement: Changed method to launch miner executables when '-ShowMinerWindows:true'; will no longer steal focus (requires [MPM-Dir]\CreateProcess.cs file)
- Enhancement: Loading pool data converted to job; shortens loop time

- Fixed 'No pools available' message when no wallet was explicitly configured in the config file
- Fixed DeviceID enumeration when main screen is connected to onboard Intel HD Graphics
- Fixed donation addresses
- Fixed error displaying Pool balances, but no wallet address / username was configured
- Fixed 'ExcludeDeviceName; before it would not exclude more than one device
- Fixed generated miner objects; pool element is of type [array]
- Fixed MPM Error message when '-MinerName' was empty
- Fixed MPM starts mining if only one miner is available
- Fixed pool balances display formatting (works with any regional setting)
- Fixed rescanning devices when reconfiguring 'DeviceName' or 'ExcludeDeviceName' while MPM is running

- Miner overview: Sort algorithms requiring benchmarking alphabetically; print either 'Benchmark pending' or 'Benchmark in progress' depending on status

- Start batch files, 'UseFastestMinerPerAlgoOnly' is now enabled by default

- Updated README: NiceHash pool configuration details

#### Miner changes

- Added AMD-JCECryptonote_v0.33b15
- Added AMD-MkxminerLyra2z_v4.20
- Added CPU-JCECryptonote_v0.33m
- Added NVIDIA-CcminerMTP_v1.1.1
- Added NVIDIA-GminerEquihash_v1.14
- Added NVIDIA-MiniZEquihash_v1.0j

- Changed config file handling of \*-XmrigCryptonight_\* miners
- Changed NVIDIA-CcminerTpruvot_v2.3; removed monero and graft algos

- Disabled some AMD miners because they need special environment variable values (which is currently not supported by MPM, see https://github.com/MultiPoolMiner/MultiPoolMiner/issues/2208#issuecomment-432696257)
  They are all in the folder 'RequireNewLancher_Amd', feel free to move them to 'MinersLegacy'. If they work that's fine :-)
  - AMD-MkxminerLyra2RE2_v3.10
  - AMD-Sgminer_v5.6.1
  - AMD-SgminerAvermore_v1.4.1
  - AMD-SgminerKL_v1.0.9
  - AMD-SgminerLyra2RE2_v5.1.0
  - AMD-SgminerXevan_v5.5.0

- Enabled AMD_NVIDIA-EminerEthash_v0.6.1rc2 dev fee

- Fixed AMD_CPU_NVIDIA-FireiceCryptonight_v2.7.1; preserve existing config files

- Fixed AMD_CPU_NVIDIA-FireiceCryptonight_v2.7.1: disabled 2 threads on NVIDIA (miner error 'NVIDIA Invalid Result GPU ID n')
- Fixed AMD-XmrigCryptonight_v2.8.6 algo filtering based on minimum memory (CryptonightHeavy requires 4GB)
- Fixed CPU-ClaymoreCryptonight_v4.0; api port is now in the 4000 range as all other miners
- Fixed CPU-CryplyAvx_v3.8.8.3 Sha256 hash
- Fixed CPU-CryplyAvx2_v3.8.8.3 Sha256 hash
- Fixed CPU-CryplyAvx2Sha_v3.8.8.3 Sha256 hash
- Fixed CPU-CryplySse42_v3.8.8.3 Sha256 hash
- Fixed CPU-CryplySse42Sha_v3.8.8.3 Sha256 hash

- Merged API files for XmRig & XmRigCfgFile

- Reported graphics memory is rounded to the nearest 1/4 GB (Include.psm1/Get-Device)

- Removed AMD_NVIDIA-lolMinerEquihash_v0.6beta4
- Removed AMD-MkxminerLyra2RE2_v3.10; it was for Lyre2RE2 only which is an unprofitable algo
- Removed AMD-SgminerLyra2RE2_v5.1.0; miner was incorrectly named, it was for Lyre2RE only which is an unprofitable algo

- Updated AMD-CastXmrVegaCryptonight_v1.66; support for algo CryptonightXfh, added support for Baffin chipset
- Updated AMD-SgminerKL_v1.0.9
- Updated AMD-SRBMinerCryptonight_v1.7.6
- Updated AMD-TeamRed_v0.3.8
- Updated AMD-WildRig_v0.15.0.12beta; support for Dedal, Lyra2v3, Lyra2vc0ban, MTP, & X21s algo
- Updated AMD-XmrigCryptonight_v2.8.6
- Updated AMD_CPU_NVIDIA-FireiceCryptonight_v2.7.1
- Updated AMD_NVIDIA-ClaymoreEthash_v12.0
- Updated AMD_NVIDIA-EthminerEthash_v0.16.2
- Updated AMD_NVIDIA-lolMinerEquihash_v0.6; support for Equihash 210/9 algo
- Updated AMD_NVIDIA-PhoenixminerEthash_v4.0b; (support for dual mining Blake2s with AMD cards is not working in 4.0b!)
- Updated NVIDIA-BMiner_v11.2.0; support for '-pers auto', algos Aeternity & BEAM, and ETH/BTM dual mine
- Updated NVIDIA-CcminerAlexis_v1.4; requires CUDA 10.0.0
- Updated NVIDIA-CcminerTrex_v0.8.9; support for Dedal, Geek, GltAstralHash, GltJeongHash, GltPadiHash, GltPawelHash & X21s algo
- Updated NVIDIA-CcminerKlausT_vYescryptR16R32.V7
- Updated NVIDIA-CcminerSuprMiner_v2.0.7; support for X16Rt
- Updated NVIDIA-CcminerTpruvot_v2.3; requires CUDA 10.0.0, support for Exosis algo
- Updated NVIDIA-CcminerZealot_v1.28; remove support for Renesis algo
- Updated NVIDIA-CryptoDredge_v0.15.2; support for algo Dedal, Lyra2vc0banHash, Lyra2V3, MTP, Pipe & X21s, requires min. 5GB of VRAM
- Updated NVIDIA-ExcavatorNHMP_v1.5.14a

#### Pool changes
- Added *Coins variants for pools that support it
- Added Ravenminer & RavenminerEU
- Added StarPool & StarPoolCoins
- Added Zpool region NA (USA)

- Fixed BlockMasters.ps1; returned empty data in some scenarios
- Fixed MiningPoolHub(Coins); changed hosts for Equihash-BTG
- Fixed NLPool; was using wrong host address (Zpool)
- Fixed PhiPhiPool URLs, added regions "EU" and "ASIA"
- Fixed Yiimp balance (message 'no wallet specified')

- Pool queries are less likely to fail (try 3 times before giving up)
- Pool objects contain fee information, will be used for profit calculation

#### New algorithms in Algorithms.txt
- Dedal
- Equihash 210/9
- Geek
- GltAstralHash
- GltJeongHash
- GltPadiHash
- GltPawelHash
- Lyra2v3
- Lyra2vc0banHash
- MTP
- Pipe
- X18
- X21s

#### Other changes / Fixes
- Get-Binaries: Fixed compatibility with miner filename change introduced in 3.1.0
- Updated README: NiceHash pool configuration details

grantemsley commented:
Why did .gitignore and .gitattributes change?  Those changes don't look right.

UselessGuru commented:
I fixed the github attributes

aaronsace commented:
Took ages to intially fix the NVidia drivers but the AMD and NVidia cards bencmarked overnight without BSOD. Will see if it looks alright this evening and have a check through your code.

UselessGuru commented:
> Took ages to intially fix the NVidia drivers

Same here.... I still don't know the exact steps to get it all working 👎

aaronsace commented:
[MultiPoolMiner-Version-3.2.0.zip](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2734589/MultiPoolMiner-Version-3.2.0.zip)

Errors: 
```
Measure-Object : The property "Earning_Bias" cannot be found in the input for any objects.
At C:\Users\aaron\Downloads\MultiPoolMiner-Version-3.2.0\MultiPoolMiner.ps1:660 char:380
+ ... ect).Count}, {([Double]($_ | Measure-Object Earning_Bias -Sum).Sum)}, ...
+                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [Measure-Object], PSArgumentException
    + FullyQualifiedErrorId : GenericMeasurePropertyNotFound,Microsoft.PowerShell.Commands.MeasureObjectCommand
```

```
New-Object : Cannot find type [Mkxminer]: verify that the assembly containing this type is loaded.
At C:\Users\aaron\Downloads\MultiPoolMiner-Version-3.2.0\MultiPoolMiner.ps1:704 char:30
+             $ActiveMiners += New-Object $Miner.API -Property @{
+                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidType: (:) [New-Object], PSArgumentException
    + FullyQualifiedErrorId : TypeNotFound,Microsoft.PowerShell.Commands.NewObjectCommand
```

aaronsace commented:
@UselessGuru This was the only thing that worked for me: 
Select the driver manually from https://www.nvidia.com/drivers
Run the installation.
Select 'Custom Install'
Tick 'Clean Install'

UselessGuru commented:
@aaronsace 

`Measure-Object : The property "Earning_Bias" cannot be found in the input for any objects.`
Fixed with https://github.com/MultiPoolMiner/MultiPoolMiner/pull/2283/commits/9df6b3eead80f7632aa2cc40629b044180074c07

`New-Object : Cannot find type [Mkxminer]: verify that the assembly containing this type is loaded.`
Fixed with https://github.com/MultiPoolMiner/MultiPoolMiner/pull/2283/commits/d0f0c20a5f9cecb1e68570d5b7c9f344d2c7fd7a

UselessGuru commented:
@aaronsace 

Any news?

aaronsace commented:
@UselessGuru I think it's running fine now; I'm just checking through logic of some of your implementations.

UselessGuru commented:
I'like to hear what implmented features you like, and where you think they could be done better / more elegantly.

As you see version 3.3.0 is waiting ;-)
https://github.com/MultiPoolMiner/MultiPoolMiner/releases/tag/Beta3.3

aaronsace commented:
The new miner files still need implementing as the current ones were meant to be deprecated before version 3 but that's not related to this PR. I'll see if I can dig out as to what was supposed to be happening; think all the pools were going to be passed to the miner files and the config file would have full control over the miner files with global and local settings. Anyway, off topic.

aaronsace commented:
@UselessGuru Why are there two functions for TCP Connections? 

https://github.com/MultiPoolMiner/MultiPoolMiner/blob/7de44451e1a080c5e1e23460941107bc325aee3a/Include.psm1#L647-L674

https://github.com/MultiPoolMiner/MultiPoolMiner/blob/7de44451e1a080c5e1e23460941107bc325aee3a/Include.psm1#L676-L716

aaronsace commented:
@UselessGuru Why are you getting the pool data in a job? 

https://github.com/MultiPoolMiner/MultiPoolMiner/blob/7de44451e1a080c5e1e23460941107bc325aee3a/MultiPoolMiner.ps1#L400-L418

UselessGuru commented:
@aaronsace 

>  function Get-TCPResponse

Invoke-TcpRequest did not work with the (now removed) old AMD_NVIDIA-lolMinerEquihash_v0.5bugfix API ('API = lolMiner). So I created function Get-TCPResponse to use with that API.

lolMiner 0.6 and later use a different API (API-lolMinerAPI) and works well with Invoke-TcpRequest

So this part of the code is obsolete and can be deleted if desired.

UselessGuru commented:
> Why are you getting the pool data in a job?

Getting the pool data can take quite some time.
By changing it to a job it can be run in the background, which reduces the overall loop time, which is especially important for benchmarking.

UselessGuru commented:
> @UselessGuru Have these calculations changed and, if so, what was the change?

No. Just formatting change in https://github.com/MultiPoolMiner/MultiPoolMiner/commit/829c68837d16a892cc38a434c178bb1adb7897f5#diff-3a447b104ea55289de9757e0588b9862

To make it more readable.

aaronsace commented:
> > Why are you getting the pool data in a job?
> 
> Getting the pool data can take quite some time.
> By changing it to a job it can be run in the background, which reduces the overall loop time, which is especially important for benchmarking.

The primary purpose of the loop is to recalculate the profits based on the pool data so there would be little reason to proceed without the calculations. 

Even during a benchmark, you get better results by waiting so if it slowed the benchmark down slightly then that wouldn't be a problem plus you have some pool stats ready for when the benchmark finishes. 

But you wait for the job to complete before continuing anyway so the job basically achieves nothing as far as I can see: 

https://github.com/MultiPoolMiner/MultiPoolMiner/blob/7de44451e1a080c5e1e23460941107bc325aee3a/MultiPoolMiner.ps1#L409

aaronsace commented:
Regarding pools and jobs, the way I imaged it working would be to batch them. If you created 4 jobs and split the pools between them then it would decrease the wait time.

aaronsace commented:
> @aaronsace
> 
> > function Get-TCPResponse
> 
> Invoke-TcpRequest did not work with the (now removed) old AMD_NVIDIA-lolMinerEquihash_v0.5bugfix API ('API = lolMiner). So I created function Get-TCPResponse to use with that API.
> 
> lolMiner 0.6 and later use a different API (API-lolMinerAPI) and works well with Invoke-TcpRequest
> 
> So this part of the code is obsolete and can be deleted if desired.

Think their API had a bug if you had to get bytes to output a string; may as well remove it if no longer needed.

UselessGuru commented:
> But you wait for the job to complete before continuing anyway

Of course so we have fresh stats. Under normal circumstances the job has just completed very shortly before the wait-job line (within a few seconds).

> so the job basically achieves nothing as far as I can see:

Not quite: 
I measure the time it took to read the stats ($GetPoolDataJobDuration.TotalSeconds)
During each loop in different places I compare the time left until the loop ends with the time needed to gather the pool data
```
        #Get pool data just in time - required if $Config.Interval is high to have recent pool data
        if (-not (Get-Job -Name GetPoolData -ErrorAction SilentlyContinue) -and (Get-Date).ToUniversalTime().AddSeconds($GetPoolDataJobDuration.TotalSeconds) -ge $StatEnd) {
            Write-Log "Triggered loading pool information. "
            $GetPoolDataJob = Start-Job -Name GetPoolData -InitializationScript ([scriptblock]::Create("Set-Location('$(Get-Location)')")) -ArgumentList ($StatSpan, $Config) -FilePath .\Get-PoolData.ps1
        }

```
and fire the pool job just in time so the stats should just about fresh for the next loop start.

This actually works quite well, as you can see in the logs, Look for these entries:


My approach shortens the time needed to benchmark, and we still get very fresh stats.

Depending on the number of active pools the polling my take up to 2 minutes. My code drastically reduces benchmark time, and does not do any harm and is fully compatible.
During benchmarking the first loop for a miner will always get extended to a full interval (to shave off the time needed to gather pool & miner info and calculations), so the stats might not be as fresh, but for benchmarking this is not really important.
For normal operations we have fresh stats like before.

UselessGuru commented:
> If you created 4 jobs and split the pools between them then it would decrease the wait time.

Agreed, that will make it faster, but not AS fast ;-)

UselessGuru commented:
> There must be something wrong with the way you are measuring the speed

I measure the time the job to gather pool data took. And I know how long the loop still has to go. -> I can start the job JIT.

> jobs don't make things faster.

I know.

But kicking off a background job while still gathering miner data reduces the wait time for new pool stats.

aaronsace commented:
Gonna have to sleep on this but I think you are effectivley gathering the pool stats whilst you are gathering the miner stats so that you gather all the stats at the same time by using line 810 instead of 406.  

https://github.com/MultiPoolMiner/MultiPoolMiner/blob/7de44451e1a080c5e1e23460941107bc325aee3a/MultiPoolMiner.ps1#L810

UselessGuru commented:
> too tired to think it though; Get-PoolData is listed 7 times in the code.
>  you are effectively gathering the pool stats whilst you are gathering the miner stats

It all depends on $Interval.

I get the pool stats **while the main loop** is running (the pool data job gets triggered depending on the remaining time of the main loop (thats why the code is there 7 times, approximatively evenly spread so I can catch the optimal stating time)

Line 810 is one of the seven instances.
I agree that having 7 possible startpoints for Get-PoolData is not ideal. But this is the best I could come up with if I want to have the freshest possible stats, while still not having to wait for them before I can start collecting the miner data.

Example 1:
Interval =  60 secs
GetPooData duration =120 seconds
MPM will start polling pool data again immediately in line 416. If the main loop will be finished before we have the pool stats I need to wait in line 409.
Note: If pool data polling is part of the main loop (original code) then there is no time at all to read the miner speed (this actually poses a problem to get a reasonable number of hash rate samples)

Example 2:
Interval =  300 secs
GetPooData duration =120 seconds
MPM wil run through the code und will check in 7 spots whether it is the right time to start GetPoolData. If the expected remaining loop time is greater that the time needed for GetPoolData, it will continue and NOT run the job.
In this scenario it will start GetPoolData towards the end (979) so that the GetPoolData job will finish JIT, so that we can immediately start getting the miner stats.

aaronsace commented:
The "Do nothing for a few seconds" 'for' loop now appears to be doing a lot of stuff but there is no guarantee that it will run. Please clarify as to what is going on inside that loop.

UselessGuru commented:
I am currently testing a version that reads the pools as batch job at the beginning of the main loop. Lokks good so far.

aaronsace commented:
Where is CpuFeatureDetector.exe compiled from?

UselessGuru commented:
> Where is CpuFeatureDetector.exe compiled from?

Grant has the source code. I think he posted about is somewhere in a close thread here.

grantemsley commented:
CpuFeatureDetector is from https://github.com/grantemsley/FeatureDetector - which is my fork of https://github.com/Mysticial/FeatureDetector, the only real change being to reformat the output so it's easier to parse.

aaronsace commented:
@grantemsley Can you change the project settings from 'application' to 'library' so that we have a 'dll' file instead of an 'exe' file? Then just pair it with a simple C# using dll imports like this but you would only have one funtion: https://github.com/MultiPoolMiner/MultiPoolMiner/blob/master/OpenCL/NativeMethods.cs

Doing this is quicker than trying to convert it to C# (if that's even possible) and it removes the 'exe' from the repository.

UselessGuru commented:
@aaronsace 

> Can you guys check it out and see if this is what we need?

Does not quite do what we need.... 
From CPU-CryplyAvx2Sha_v3.8.8.3.ps1:
` $Devices = $Devices | Where-Object Type -EQ "CPU" | Where-Object {(-not $_.CpuFeatures) -or ($_.CpuFeatures -contains "avx2" -and $_.CpuFeatures -contains "sha")}`

Your code dos not expose AVX2 ...

aaronsace commented:
Take 2. 

There was also an asembly line (`mov ecx, 0`) missing which made it highley confusing. 

It should now (hopefully) match the exe exactly: 

```
using namespace System.Reflection
using namespace System.Reflection.Emit
using namespace System.Runtime.InteropServices

Add-Type @"
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Reflection;
using System.Runtime.InteropServices;

public static class CpuID {

    public static byte[] Invoke(int level) {
        IntPtr codePointer = IntPtr.Zero;
        try {
            // compile
            byte[] codeBytes;
            if (IntPtr.Size == 4) {
                codeBytes = x86CodeBytes;
            } else {
                codeBytes = x64CodeBytes;
            }

            codePointer = VirtualAlloc(
                IntPtr.Zero,
                new UIntPtr((uint)codeBytes.Length),
                AllocationType.COMMIT | AllocationType.RESERVE,
                MemoryProtection.EXECUTE_READWRITE
            );

            Marshal.Copy(codeBytes, 0, codePointer, codeBytes.Length);

            CpuIDDelegate cpuIdDelg = (CpuIDDelegate)Marshal.GetDelegateForFunctionPointer(codePointer, typeof(CpuIDDelegate));

            // invoke
            GCHandle handle = default(GCHandle);
            var buffer = new byte[16];

            try {
                handle = GCHandle.Alloc(buffer, GCHandleType.Pinned);
                cpuIdDelg(level, buffer);
            } finally {
                if (handle != default(GCHandle)) {
                    handle.Free();
                }
            }

            return buffer;
        } finally {
            if (codePointer != IntPtr.Zero) {
                VirtualFree(codePointer, 0, 0x8000);
                codePointer = IntPtr.Zero;
            }
        }
    }

    [UnmanagedFunctionPointerAttribute(CallingConvention.Cdecl)]
    private delegate void CpuIDDelegate(int level, byte[] buffer);

    [DllImport("kernel32.dll", SetLastError = true)]
    private static extern IntPtr VirtualAlloc(IntPtr lpAddress, UIntPtr dwSize, AllocationType flAllocationType,
        MemoryProtection flProtect);

    [DllImport("kernel32")]
    private static extern bool VirtualFree(IntPtr lpAddress, UInt32 dwSize, UInt32 dwFreeType);

    [Flags()]
    private enum AllocationType : uint {
        COMMIT = 0x1000,
        RESERVE = 0x2000,
        RESET = 0x80000,
        LARGE_PAGES = 0x20000000,
        PHYSICAL = 0x400000,
        TOP_DOWN = 0x100000,
        WRITE_WATCH = 0x200000
    }

    [Flags()]
    private enum MemoryProtection : uint {
        EXECUTE = 0x10,
        EXECUTE_READ = 0x20,
        EXECUTE_READWRITE = 0x40,
        EXECUTE_WRITECOPY = 0x80,
        NOACCESS = 0x01,
        READONLY = 0x02,
        READWRITE = 0x04,
        WRITECOPY = 0x08,
        GUARD_Modifierflag = 0x100,
        NOCACHE_Modifierflag = 0x200,
        WRITECOMBINE_Modifierflag = 0x400
    }

    // Basic ASM strategy --
    // void x86CpuId(int level, byte* buffer) 
    // {
    //    eax = level
    //    cpuid
    //    buffer[0] = eax
    //    buffer[4] = ebx
    //    buffer[8] = ecx
    //    buffer[12] = edx
    // }

    private readonly static byte[] x86CodeBytes = {
        0x55,                   // push        ebp  
        0x8B, 0xEC,             // mov         ebp,esp
        0x53,                   // push        ebx  
        0x57,                   // push        edi

        0x8B, 0x45, 0x08,       // mov         eax, dword ptr [ebp+8] (move level into eax)
        0x0F, 0xA2,              // cpuid

        0x8B, 0x7D, 0x0C,       // mov         edi, dword ptr [ebp+12] (move address of buffer into edi)
        0x89, 0x07,             // mov         dword ptr [edi+0], eax  (write eax, ... to buffer)
        0x89, 0x5F, 0x04,       // mov         dword ptr [edi+4], ebx 
        0x89, 0x4F, 0x08,       // mov         dword ptr [edi+8], ecx 
        0x89, 0x57, 0x0C,       // mov         dword ptr [edi+12],edx 

        0x5F,                   // pop         edi  
        0x5B,                   // pop         ebx  
        0x8B, 0xE5,             // mov         esp,ebp  
        0x5D,                   // pop         ebp 
        0xc3                    // ret
    };

    private readonly static byte[] x64CodeBytes = {
        0x53,                         // push rbx    this gets clobbered by cpuid
    
        // rcx is level
        // rdx is buffer.
        // Need to save buffer elsewhere, cpuid overwrites rdx
        // Put buffer in r8, use r8 to reference buffer later.        
    
        // Save rdx (buffer addy) to r8
        0x49, 0x89, 0xd0,             // mov r8,  rdx
    
        // Move ecx (level) to eax to call cpuid, call cpuid
        0x89, 0xc8,                   // mov eax, ecx
        0xB9, 0x00, 0x00, 0x00, 0x00, // mov ecx, 0
        0x0F, 0xA2,                   // cpuid
    
        // Write eax et al to buffer
        0x41, 0x89, 0x40, 0x00,       // mov    dword ptr [r8+0],  eax
        0x41, 0x89, 0x58, 0x04,       // mov    dword ptr [r8+4],  ebx
        0x41, 0x89, 0x48, 0x08,       // mov    dword ptr [r8+8],  ecx
        0x41, 0x89, 0x50, 0x0c,       // mov    dword ptr [r8+12], edx
    
        0x5b,                         // pop rbx
        0xc3                          // ret
    };
}
"@

# Brief : gets CPUID (CPU name and registers)
function Get-CpuId {
    #OS Features
    $OS_x64 = "" #not implemented
    $OS_AVX = "" #not implemented
    $OS_AVX512 = "" #not implemented

    #Vendor
    $vendor = "" #not implemented

    if ($vendor -eq "GenuineIntel") {
        $Vendor_Intel = $true;
    }
    elseif ($vendor -eq "AuthenticAMD") {
        $Vendor_AMD = $true;
    }

    $info = [CpuID]::Invoke(0)
    #convert 16 bytes to 4 ints for compatibility with existing code
    $info = [int[]]@(
        [BitConverter]::ToInt32($info, 0*4)
        [BitConverter]::ToInt32($info, 1*4)
        [BitConverter]::ToInt32($info, 2*4)
        [BitConverter]::ToInt32($info, 3*4)
    )

    $nIds = $info[0]

    $info = [CpuID]::Invoke(0x80000000)
    $nExIds = [BitConverter]::ToUInt32($info, 0*4) #not sure as to why 'nExIds' is unsigned; may not be necessary
    #convert 16 bytes to 4 ints for compatibility with existing code
    $info = [int[]]@(
        [BitConverter]::ToInt32($info, 0*4)
        [BitConverter]::ToInt32($info, 1*4)
        [BitConverter]::ToInt32($info, 2*4)
        [BitConverter]::ToInt32($info, 3*4)
    )

    #Detect Features
    $features = @{}
    if ($nIds -ge 0x00000001) {

        $info = [CpuID]::Invoke(0x00000001)
        #convert 16 bytes to 4 ints for compatibility with existing code
        $info = [int[]]@(
            [BitConverter]::ToInt32($info, 0*4)
            [BitConverter]::ToInt32($info, 1*4)
            [BitConverter]::ToInt32($info, 2*4)
            [BitConverter]::ToInt32($info, 3*4)
        )

        $features.MMX    = ($info[3] -band ([int]1 -shl 23)) -ne 0
        $features.SSE    = ($info[3] -band ([int]1 -shl 25)) -ne 0
        $features.SSE2   = ($info[3] -band ([int]1 -shl 26)) -ne 0
        $features.SSE3   = ($info[2] -band ([int]1 -shl 00)) -ne 0

        $features.SSSE3  = ($info[2] -band ([int]1 -shl 09)) -ne 0
        $features.SSE41  = ($info[2] -band ([int]1 -shl 19)) -ne 0
        $features.SSE42  = ($info[2] -band ([int]1 -shl 20)) -ne 0
        $features.AES    = ($info[2] -band ([int]1 -shl 25)) -ne 0

        $features.AVX    = ($info[2] -band ([int]1 -shl 28)) -ne 0
        $features.FMA3   = ($info[2] -band ([int]1 -shl 12)) -ne 0

        $features.RDRAND = ($info[2] -band ([int]1 -shl 30)) -ne 0
    }

    if ($nIds -ge 0x00000007) {

        $info = [CpuID]::Invoke(0x00000007)
        #convert 16 bytes to 4 ints for compatibility with existing code
        $info = [int[]]@(
            [BitConverter]::ToInt32($info, 0*4)
            [BitConverter]::ToInt32($info, 1*4)
            [BitConverter]::ToInt32($info, 2*4)
            [BitConverter]::ToInt32($info, 3*4)
        )

        $features.AVX2         = ($info[1] -band ([int]1 -shl 05)) -ne 0

        $features.BMI1         = ($info[1] -band ([int]1 -shl 03)) -ne 0
        $features.BMI2         = ($info[1] -band ([int]1 -shl 08)) -ne 0
        $features.ADX          = ($info[1] -band ([int]1 -shl 19)) -ne 0
        $features.MPX          = ($info[1] -band ([int]1 -shl 14)) -ne 0
        $features.SHA          = ($info[1] -band ([int]1 -shl 29)) -ne 0
        $features.PREFETCHWT1  = ($info[2] -band ([int]1 -shl 00)) -ne 0

        $features.AVX512_F     = ($info[1] -band ([int]1 -shl 16)) -ne 0
        $features.AVX512_CD    = ($info[1] -band ([int]1 -shl 28)) -ne 0
        $features.AVX512_PF    = ($info[1] -band ([int]1 -shl 26)) -ne 0
        $features.AVX512_ER    = ($info[1] -band ([int]1 -shl 27)) -ne 0
        $features.AVX512_VL    = ($info[1] -band ([int]1 -shl 31)) -ne 0
        $features.AVX512_BW    = ($info[1] -band ([int]1 -shl 30)) -ne 0
        $features.AVX512_DQ    = ($info[1] -band ([int]1 -shl 17)) -ne 0
        $features.AVX512_IFMA  = ($info[1] -band ([int]1 -shl 21)) -ne 0
        $features.AVX512_VBMI  = ($info[2] -band ([int]1 -shl 01)) -ne 0
    }

    if ($nExIds -ge 0x80000001) {

        $info = [CpuID]::Invoke(0x80000001)
        #convert 16 bytes to 4 ints for compatibility with existing code
        $info = [int[]]@(
            [BitConverter]::ToInt32($info, 0*4)
            [BitConverter]::ToInt32($info, 1*4)
            [BitConverter]::ToInt32($info, 2*4)
            [BitConverter]::ToInt32($info, 3*4)
        )

        $features.x64   = ($info[3] -band ([int]1 -shl 29)) -ne 0
        $features.ABM   = ($info[2] -band ([int]1 -shl 05)) -ne 0
        $features.SSE4a = ($info[2] -band ([int]1 -shl 06)) -ne 0
        $features.FMA4  = ($info[2] -band ([int]1 -shl 16)) -ne 0
        $features.XOP   = ($info[2] -band ([int]1 -shl 11)) -ne 0
    }

    # wrap data into PSObject
    New-Object PSObject -Property @{
        Vendor   = $vendor
        Name     = $name
        Features = $features.Keys.ForEach{if ($features.$_) {$_}}
    }
}

(Get-CpuId).Features
```

aaronsace commented:
The way sleep time is calculated on this seems too complicated and doesn't seem to be following the interval: 
https://github.com/MultiPoolMiner/MultiPoolMiner/commit/1efbafbdd446df43737b4adfb7f7beaa5eab7040

`-and` meant that it would run a maximum of 10 times
`-or` means that it will run a minimum of 10 times

Shouldn't this be something like 10 whilst benchmarking and 1 whilst running?

UselessGuru commented:
> Shouldn't this be something like 10 whilst benchmarking and 1 whilst running? Or just vary between 10 and 1 depending on how much time is available?

I am not happy with that code too... It does not work well with long intervals (we need 2 intervals: $Interval & $BenchmarkInterval)
I am working on something better... stay tuned

aaronsace commented:
@UselessGuru I'm going to strip things down a bit. I think a lot of additions appear to have been added to fix problems that should never have been problems to begin with. Apologies for delaying this pull request longer than expected.

UselessGuru commented:
>  I think a lot of additions appear to have been added to fix problems that should never have been problems to begin with

I don't understand. The old code does not work well with very short or very long intervals.
My code changes fixes all that.

aaronsace commented:
@UselessGuru Please don't push any more commits to MultiPoolMiner.ps1

ChromaticProductions commented:
@Aaronsace A big thank you and the dev team for MPM.  Is the final release of v3.2.0 still pending the rewrite of the code ‘bit related to hashrates’ or did we skip the final release and jump to v3.3.0 development?  Thank you for the great product.

aaronsace commented:
@ChromaticProductions I know I've already apologised before but I'm trying to get the updates to MultiPoolMiner.ps1 ASAP which has grown into a lot bigger task than anticipated. It's taking some time working through a lot of the bloat that's built up and not break the functionality but, once this is branch is merged into master, the master will be merged up to the 3.3 branch.

UselessGuru commented:
@aaronsace 

I am not happy with some of the rewritten stuff, especially the hash rate gathering stuff.
Your code has issues with long intervals when benchmarking, and it does not shorten the benchmarking to a mininum time.
Also the 'kill stray miner' code does not work with orphaned Excavator processes.

Other changes are perfect, e.g. 
- Get-ChildContent using jobs
- creating Config file
- balances stuff

I have merged a good part of your code into 3.3.0 Beta 8.

aaronsace commented:
@UselessGuru 

This is causing hanging/locking issues with the pools: 
https://github.com/MultiPoolMiner/MultiPoolMiner/blob/aa288113c8a2221a7be6199beab8647d274d29d3/MultiPoolMiner.ps1#L341-L344

So, please move `#Load the stats` to just before `#Load information about the miners`: 
https://github.com/MultiPoolMiner/MultiPoolMiner/blob/aa288113c8a2221a7be6199beab8647d274d29d3/MultiPoolMiner.ps1#L430

aaronsace commented:
@UselessGuru 

The `GetHashRate` function handles everything; the core code doesn't need to do anything in regards to benchmarks except call it. No defined times are needed; the benchmark ends as soon as the 'GetHashRate' function returns (or repeatedly doesn't return) a value. 

Please check this line because it appears to run way longer then 10 seconds sometimes: 
https://github.com/MultiPoolMiner/MultiPoolMiner/blob/aa288113c8a2221a7be6199beab8647d274d29d3/MultiPoolMiner.ps1#L295

aaronsace commented:
@UselessGuru 

I think all the log reader commands have got messed up; please amend to: 
`start pwsh -noexit -executionpolicy bypass -command "& .\reader.ps1 -log 'MultiPoolMiner_\d\d\d\d-\d\d-\d\d\.txt' -sort '^[^_]*_' -quickstart"`

UselessGuru commented:
> Solution cannot depend on file location: `"$(Get-Location)\Bin\*"`

Why not? This is the only way you can kill any miner processes from a previously crashed MPM run.

> stray services could be tackled after release as to not delay.

I will not revert it. Feel free to do with the code what ever you do, but I'd appreciated you discuss and question stuff before just overriding valuable work.

> start pwsh -noexit -executionpolicy bypass -command "& .\reader.ps1 -log 'MultiPoolMiner_\d\d\d\d-\d\d-\d\d\.txt' -sort '^[^_]*_' -quickstart"

Done in https://github.com/MultiPoolMiner/MultiPoolMiner/pull/2283/commits/705834fbfac1681f79f695bb62b269e86254ba54

> Not sure of the correct implementation, please check. Return the array (`[Int64[]]`)? Or sum the values (`Measure-Object -Sum`)?

You are returning only one algo, so summing it up seem more correct. 
Hashrates should be of type [double] because some algos produce very low values, e.g. Equihash 1505.
Changed in https://github.com/MultiPoolMiner/MultiPoolMiner/pull/2283/commits/5e7533ddb02da3c27b69436061819133eecfcaa5

> Please check this line because it appears to run way longer than 10 seconds sometimes:

Yes, it does take longer when MPM is running other tasks in the background. Probably caused by 
#Set process priority to BelowNormal to avoid hash rate drops on systems with weak CPUs
    $ConfigBackup = $Config | ConvertTo-Json -Depth 10 | ConvertFrom-Json	(Get-Process -Id $PID).PriorityClass = "BelowNormal"

> This is causing hanging/locking issues with the pools:

Stats must be read before the pools (been like this since the very beginning). I don't understand why there would be a lock.

aaronsace commented:
A `Miner` has control over everything including what files should/shouldn't be running. The bin folder is abritory; a miner could be running files from Program Files, a network share or anywhere. If GitHub can't revert then just override. 

Seems more the a prority issue; there is a huge discrency sometimes beyond normal. If I get chance, I will try and issolate what's wrong if no issues are appearing for you. Have a test on Saturday so I'm meant to be revising for that at the moment...

Pools is writting hundreds of files whilst `Get-Stat` is reading the folder. First use of `$Stats` is at `#Load information about the miners` to save miners time over calling `Get-Stat` 1000s of times. It just needs moving to there.

UselessGuru commented:
> Pools is writting hundreds of files whilst

Pools are reading  hundreds of files too:

 if ((Get-Stat -Name "$($Name)_$($Algorithm_Norm)_Profit") -eq $null)

aaronsace commented:
That's on the same thread, it won't concurrently read and write the same file at the same time. Apart from a minor speed decrease from reading a file before writing to it, it is benign.

UselessGuru commented:
> The `GetHashRate` function handles everything

There is an issue with using a time window.
Using a defined number of samples is better because it allows to use data from previous runs (which might have expired otherwise). Especially alogithms with large fluctuation, e.g. X16 profit from keeping a large number of samples.
Alternatively we could ALWAYS use ALL samples, but this would slow down everything after a while. So limiting the samples to something like 100 x $Miner.BenchmarkInterval sounds reasonable.

aaronsace commented:
It handles that with the `GetHashRate` `Seconds` parameter in combination with the new `Miner` `IntervalMultplier` variable. The core will even call `GetHashRate` multiple times with various `Seconds` values in an effort to get a valid reading.

UselessGuru commented:
> It handles that with the `GetHashRate` `Seconds` parameter

That is where the issue comes from:

Algos like X16 require loads of samples for stable hashrates. The more the better. The time window in your code does not allow old samples, e.g. from an previous run 2hrs ago, so become part of the calculatiation. This is why the sample count is better than the time frame.

aaronsace commented:
@UselessGuru Increase the `IntervalMultplier` if you are worried about the x16 hashrate but the code accounts for hashrates from the past 168 hours regardless of the algorithm.

aaronsace commented:
What may be helpful, if absolute necessary, is to enforce a minimum `Stat` file `Duration` to which a miner is considered to have finished benchmarking. Instead of considering it complete just because it has a `Stat` file.

UselessGuru commented:
> but the code accounts for hashrates from the past 168 hours regardless of the algorithm

```
...
                $Miner_Algorithm = $_ -replace "-NHMP" #temp fix
                $Miner.Speed_Live += $Miner.GetHashRate($Miner_Algorithm, ($Config.Interval * $Miner.IntervalMultplier * 2), $false)

                $Miner_Speed = $Miner.GetHashRate($Miner_Algorithm, ($Config.Interval * $Miner.IntervalMultplier * 2), $Miner.New)
...
```

How is that 168hrs?

UselessGuru commented:
> What may be helpful, if absolute necessary, is to enforce a minimum `Stat` file `Duration`

My code does exactly this, but not not with time, but sample counts (parameter MinHashRateSamples)

aaronsace commented:
`MinHashRateSamples` is the equivalent of `IntervalMultplier` but `MinHashRateSamples` can be skewed by time; increasing `IntervalMultplier` is more effective.

UselessGuru commented:
@aaronsace 

> `MinHashRateSamples` is the equivalent of `IntervalMultplier`

It is not. There are situations where you will not reach a reasonable number of benchmark samples (should be at least 10) in a 'normal' benchmark loop. Your code will need to do a second loop to finish the benchmark.

As I said I will not commit this one because I disagree with some changes you made.
In my opinion your code lacks the following impreovements over my code:
- Slower benchmarking (requiring more loops)
- No fast detection of failed miners
- sub-optimal hash rate span (as discussed above)
- less useful screen output

This is your repository, you may do with the code as you please.

I have learnt a lot from you and I am happy to contribute new stuff, but it is more satisfying if you considered my suggestions and improvements in a more positive way.

I will continue developing 3.3.0 based on my code.

aaronsace commented:
@UselessGuru I think you are misreading various parts of the code (or I've made a mistake somewhere) but I will be back after Saturday to follow this up if there are still pending concerns/issues.

UselessGuru commented on MultiPoolMiner.ps1:
```diff
@@ -75,31 +78,25 @@ param(
-    [Switch]$ShowPoolBalancesExcludedPools = $false,
-    [Parameter(Mandatory = $false)]
-    [Switch]$ShowPoolBalancesDetails = $false,
+    [Alias("IgnoreFees")]
```
This should be left as 'IgnoreFees'

Version 3.3.0 will include power cost calculations and will introduce parameter 'IgnorePowerCosts'

Miner fees and power costs are not the same.

UselessGuru commented on MultiPoolMiner.ps1:
```diff
@@ -712,7 +712,7 @@ while (-not $API.Stop) {
     if ($Downloader) {$Downloader | Receive-Job}
     Start-Sleep $Config.Delay #Wait to prevent BSOD
-    Get-CIMInstance CIM_Process | Where-Object ExecutablePath | Where-Object {$_.ExecutablePath -like "$(Get-Location)\Bin\*"} | Where-Object {$ActiveMiners.ProcessID -notcontains $_.ProcessID} | Select-Object -ExpandProperty ProcessID | ForEach-Object {Stop-Process -Id $_ -Force -ErrorAction Ignore} #Kill stray miners
+    if ($ActiveMiners | ForEach-Object {$_.GetProcessNames()}) {Get-Process -Name @($ActiveMiners | ForEach-Object {$_.GetProcessNames()} | Select-Object) -ErrorAction Ignore | Select-Object -ExpandProperty ProcessName | Compare-Object @($ActiveMiners | Where-Object Best | Where-Object {$_.GetStatus() -eq "Running"} | ForEach-Object {$_.GetProcessNames()} | Select-Object) | Where-Object SideIndic ...
```
This does NOT kill any miners that might still be running from previous runs, or when MPM run stated twice from the same directory.
I don't understand why you are so opposed to my solution.

aaronsace replied:
The 'bin' folder is not a folder used by MultiPoolMiner. There are two lists being compared; one from the miners and one from MultiPoolMiner. Sounds like there is probably a bug in one of the lists.

UselessGuru replied:
>> The 'bin' folder is not a folder used by MultiPoolMiner.

Yes it is:
$_.ExecutablePath -like "$(Get-Location)\Bin\*" ponts to the MPM\Bin folder.

aaronsace replied:
The 'Bin' folder is definitely nowhere within `MultiPoolMiner.ps1` or `Include.psm1`. 

I'll have a look at this soon though as something must be wrong here.

grantemsley replied:
@UselessGuru I think what @aaronsace is *trying* to say is that the binaries are currently downloaded to .\Bin, but they don't *have* to be, and he maybe intents to change that in the future.  Please correct me if I'm wrong there.

> A `Miner` has control over everything including what files should/shouldn't be running. The bin folder is abritory; a miner could be running files from Program Files, a network share or anywhere. 

The problem is, the Miner object *should* reflect what miners, if any, are currently running, but reality doesn't always match what the objects say.  There can be miners still running from:

- a previous run of MPM that crashed or otherwise didn't clean up when it exited
- making changes to the miner files without restarting MPM
- a miner being stopped on the last cycle, but windows not cooperating, saying it was killed, and it still running.  Windows refusing to actually close a program when you tell it to isn't a code problem, it's a windows problem.  Everyone has at some point had a window where they could click the X all they wanted and nothing happens.  Sometimes even trying to kill apps from task manager fails.  It's just a windows thing.  Some of the miners take awhile to actually exit, and don't actually finish exiting on the first try, especially on systems with slow CPUs.


So the solution we put in way back when was to look at the Miner objects, look at the running processes, and try to kill any executable running in the .\Bin folder that shouldn't be running according to the Miner objects.

This brings reality back into line with what the Miner objects think is happening.

Limiting to killing executables in the .\Bin folder was to prevent it from interfering with any other executables - say if you had other miners installed elsewhere you ran along side MPM.

One possible alternative that doesn't rely on them being in the .\Bin folder would be to build a list of all the full paths of the miner executables, comparing THAT list to the "expected to be running" miners, and killing any that shouldn't be running.  Changing the miner definition files without restarting MPM could still leave something running, and it would probably be a little slower than just brute force killing whatever is in the .\Bin folder, but it would solve the other two ways I know of miners being accidentally left running.

UselessGuru replied:
> One possible alternative that doesn't rely on them being in the .\Bin folder would be to build a list of all the full paths of the miner executables,

To fetch ALL miner binary paths a separate Get-ChildItemContent with all HW activated would be needed, otherwise we might build an incomplete list that could be missing miners that have been disabled since the last run.

I'd stick with the Bin\ path version as long as all miners are using that path.

UselessGuru commented on MultiPoolMiner.ps1:
```diff
@@ -126,185 +122,109 @@ $DecayBase = 1 - 0.1 #decimal percentage
-$RunningMiners = @()
-
-$Rates = [PSCustomObject]@{BTC = [Double]1}
+$AllMinerPaths = @()
```
List of all miners paths ever active in the current run

UselessGuru commented on MultiPoolMiner.ps1:
```diff
@@ -313,154 +233,226 @@ while ($true) {
+            #Strip Model information from devices -> will create only one miner instance
+            if ($Config.CreateMinerInstancePerDeviceModel) {$DevicesTmp = $Devices} else {$DevicesTmp = $Devices | ConvertTo-Json -Depth 10 | ConvertFrom-Json; $DevicesTmp | ForEach-Object {$_.Model = ""}}
+            Get-ChildItemContent "MinersLegacy" -Parameters @{Pools = $Pools; Stats = $Stats; Config = $Config; Devices = $DevicesTmp} | 
+                ForEach-Object {$_.Content | Add-Member Name $_.Name -PassThru -Force; $_.Content.Path = $ExecutionContext.SessionState.Path.GetUnresolvedProviderPathFromPSPath($_.Content.Path); $AllMinerPaths += $_.Content.Path} | 
```
Expand path here, because $AllMinerPaths requires full paths.

UselessGuru commented on MultiPoolMiner.ps1:
```diff
@@ -537,23 +528,29 @@ while ($true) {
 
         $Miner | Add-Member DeviceName @($Miner.DeviceName | Select-Object -Unique | Sort-Object) -Force
 
-        $Miner.Path = $ExecutionContext.SessionState.Path.GetUnresolvedProviderPathFromPSPath($Miner.Path)
```
Already done in line 443

UselessGuru commented on MultiPoolMiner.ps1:
```diff
@@ -715,21 +710,18 @@ while ($true) {
-    if ($Downloader) {$Downloader | Receive-Job}
+    $API.WatchdogTimers = $WatchdogTimers #Give API access to WatchdogTimers information
     Start-Sleep $Config.Delay #Wait to prevent BSOD
+    Get-CIMInstance CIM_Process | Where-Object ExecutablePath | Where-Object {$AllMinerPaths -contains $_.ExecutablePath} | Where-Object {$ActiveMiners.ProcessID -notcontains $_.ProcessID} | Select-Object -ExpandProperty ProcessID | ForEach-Object {Stop-Process -Id $_ -Force -ErrorAction Ignore}
```
Kill all stray miners that have a matching path.
